### PR TITLE
Added file upload support to lambda. Supports multipart. Includes tests.

### LINKED ISSUE #1419: File uploads using apollo-server-lambda
Hi,
Could not find any mention, whether apollo-server-lambda supports the GraphQL multipart request specification for uploading files as mutation arguments (like apollo-server does)

### PR DESCRIPTION
**Pull request by k00k**
Fixes #1419, #1703 file upload using lambda.
Based on original [PR #1739](https://github.com/apollographql/apollo-server/pull/1739)

TODO:

- [x]  Update CHANGELOG.md with your change (include reference to issue & this PR)
- [x]  Make sure all of the significant new logic is covered by tests
- [x]  Rebase your changes on master so that they can be merged easily
- [x]  Make sure all tests and linter rules pass

After needing `graphql-upload` support in `apollo-server-lambda`, we (me and @smschick) found [PR #1739](https://github.com/apollographql/apollo-server/pull/1739) that was created a while ago and never was updated after @abernix had some comments on the PR. The main thing that @abernix had mentioned was being sure that the file upload process was [happening in the right order](https://github.com/apollographql/apollo-server/pull/1739/files#r288265931).

This PR takes those comments and some of the work from the original PR and hopefully solves the issue.  Thanks to @charleswong28 for his work on the orig PR.

Please let me know if there's anything you'd like to see. Currently all tests are passing which includes single and multiple file uploads.

### CODE REVIEW

File: packages/apollo-server-lambda/src/lambdaApollo.ts [hunk 1]
```diff
@@ -15,12 +18,13 @@ export interface LambdaGraphQLOptionsFunction {
...
+  serverParams: any
 ): lambda.APIGatewayProxyHandler {
   if (!options) {
     throw new Error('Apollo Server requires options.');
   }
 
-  if (arguments.length > 1) {
+  if (arguments.length > 2) {
     throw new Error(
       `Apollo Server expects exactly one argument, got ${arguments.length}`,
```
Comment (smschick) [inline packages/apollo-server-lambda/src/lambdaApollo.ts]: `Apollo Server expects exactly two arguments, got ${arguments.length}`,

**Comment by k00k** [comment]
Not sure why the netlify deploy preview all of a sudden started failing, all I updated were some comments. I think it's a netlify issue. Can someone please run the CI again?

**Comment by jacky-ttt** [comment]
@k00k I tried your pull request. I can upload image to s3 via `apollo-server-lambda` locally using `serverless-s3-local`. But when I upload the code to a real aws lambda. The file saved cannot be read correctly. Maybe the encoding is wrong. Have you encounter such problem? Do I need to config something on AWS Lambda?

**Comment by k00k** [comment]
@jacky-ttt I'll take a look. In the meantime, can you share your code?

**Comment by k00k** [comment]
@jacky-ttt After some testing, I believe I was able to replicate the problem you were having. This is a guess, of course, because I don't know exactly all of the steps you did or didn't take. But I was able to upload locally and get "corrupted" files when sent through a real Lambda.

I'm guessing it's likely due to needing to set API Gateway's Binary Media Types. In order to pass binary data via `multipart/form-data`, you need to add that as a _Binary Media Type_ in your API Gateway's settings for that lambda.

I've published an [example repository](https://github.com/abcya/apollo-server-lambda-example) that takes care of this for you if you deploy with the [Serverless Framework](https://serverless.com/). Otherwise, the README outlines how to do it manually.

Please let me know if that does in fact solve your problem.

If anyone else has other/better ideas on how to handle that, please feel free to chime in. But I believe it's likely out of the scope of this repo and thus shouldn't affect the merging of this PR.

**Comment by jacky-ttt** [comment]
@k00k Thanks for the quick reply. Sorry for not share some code in the first place. The example repo is really helpful. Your suspicion is correct. What I miss is to use `serverless-apigw-binary` to add binary media type, `multipart/form-data`.

**Comment by aymericbouzy** [comment]
@abernix I'm very much interested in this feature. I was wondering how much time is usually needed for such a PR to be validated and released? Thanks in advance. And thank you so much for everything that is being done on Apollo Server ❤️

**Comment by k00k** [comment]
@abernix and @martijnwalraven is there anything we can do to help get this PR moving forward?

**Comment by mrfoster** [comment]
I am still hitting the following error 
```
This implementation of ApolloServer does not support file uploads because the environment cannot accept multi-part forms
```
Thrown here: https://github.com/abcya/apollo-server/blob/fb60598286e9f45e55e3ccb197f1ce11b337014a/packages/apollo-server-core/src/ApolloServer.ts#L280

Which is caused by supportUploads always returning false in the superclass: https://github.com/abcya/apollo-server/blob/fb60598286e9f45e55e3ccb197f1ce11b337014a/packages/apollo-server-core/src/ApolloServer.ts#L670

It looks like it should be as simple as adding the following in the ApolloServer subclass.
```
  supportsUploads(): boolean {
    return true;
  }
```

I mentioned it before here:
https://github.com/apollographql/apollo-server/pull/1739#issuecomment-518531156

I'm bit puzzled how it could be working at all unless it's still being executed against a very old version.

**Comment by k00k** [comment]
@mrfoster 

You're correct, I've added the `supportsUpload` to the subclass. Good catch. 

As to why I and a few others (and the automated tests) didn't have trouble with this originally, I agree, very weird. When you tested, were you supplying an `uploads` property/obj when instantiating ApolloServer? What version of Node were you testing on? I was mainly testing with 12.14.1 (current Lambda latest).

In any case, please pull and give it a shot now, should be fine.

**Comment by mrfoster** [comment]
@k00k Thanks for adding that.

Yes, uploads option is being supplied to set maxFileSize. Using node v10.16.3 with typescript 3.7.4. 

I am actually copying apollo-server-lambda files into my project to get it working because I ran into some issues building it.  Very much looking forward to using a npm package once it's been accepted.

Thanks to you and @charleswong28 for sorting it, much appreciated.

**Comment by aymericbouzy** [comment]
@k00k @mrfoster I would like to try this PR even before it gets merged. What are the steps required to test this branch in my project?

**Comment by k00k** [comment]
> @k00k @mrfoster I would like to try this PR even before it gets merged. What are the steps required to test this branch in my project?

@aymericbouzy you should be able to do the following. There is a chance you'll have some dependency issues regarding `graphql`, but give this a shot and see how you make out.

```
git clone https://github.com/abcya/apollo-server
cd apollo-server
npm install
cd packages/apollo-server-lambda/
npm link
```
Install your local Apollo Server in your App
```
cd ~/myApp
npm link apollo-server-lambda
```

**Comment by abernix** [comment]
> This PR is on my radar (to review and hopefully release), but currently behind a number of other things (sorry!) but I'm jumping in real quick to help with @aymericbouzy's question on how to test it before it's released: 

**However!** You should be able to test it without `npm link`-ing it by using the package tarballs, which are available in the CircleCI job for _Package tarballs_ which are available via the _Details_ link on the GitHub status checks section on this PR (i.e. if you expand where it says _7 successful checks_).  Once there, you can click on the _Artifacts_ tab of CircleCI and see the available tarballs.  ([Direct link to that page](https://circleci.com/gh/apollographql/apollo-server/59244#artifacts/containers/0))

In other words, you should be able to run this to install it from the tarball directly:

```
npm install https://59244-56760633-gh.circle-artifacts.com/0/packed/apollo-server-lambda/apollo-server-lambda-2.9.16.tgz
```

(This same pattern should work for any PR opened on the Apollo Server repository.)

Please do provide feedback on your success using it, as nothing beats real-world usage in terms of landing features. 😄

**Comment by aymericbouzy** [comment]
Thanks @abernix! In fact I _don't_ see the _Artifacts_ tab ... but your direct link works! Seems to be a weird bug of CircleCI.

Screenshot 2020-01-31 at 16 35 38

> npm install https://59244-56760633-gh.circle-artifacts.com/0/packed/apollo-server-lambda/apollo-server-lambda-2.9.16.tgz

Will try this way and provide feedback 👍

**Comment by abernix** [comment]
Interesting!  Thanks for the note, @aymericbouzy.

That is indeed peculiar/buggy, more so since it shows the artifact tree _without_ the tab when I go to that link in _Incognito_ mode.  It looks like the tab reliably shows, so long as you're logged in as SOME user in their "New UX" experience: https://app.circleci.com/jobs/github/apollographql/apollo-server/59244/artifacts

Looking forward to knowing how your use of it goes!  (No hurry.)

**Comment by aymericbouzy** [comment]
I managed to make it work 🎉 

Some feedback on my "developer experience" : 

1. using the artifact was a breeze 👏 

2. I'm using [Serverless Framework](https://serverless.com/) : I had to add the following configuration to make it work (with provider `aws`): 

```yaml
provider:
  apiGateway:
    binaryMediaTypes:
      - "*/*"
```

I thought this was not very well documented, so including it in the documentation will help people get started I believe. What's particularly hard about this bug is that when this property is missing, the file simply shows up as corrupted, but there are no errors per say, no warning, nothing to point you at what is wrong about your setup. It is not defined this way by default, which may be an issue with Serverless Framework? I don't know why it is this way, but at the moment, this is a real footgun.

3. Since this feature hides away the implementation details and the file is provided as a stream, it might not be clear for the user that the file is actually not _streamed_ into the AWS Lambda : the handler receives the whole file in the event, and the file is simply provided as a stream to be consumed. But this is not memory efficient in any way (unless I totally misunderstand this) and configuring the limits in terms of file size seems important since AWS Lambda runs in a limited memory environment. However, the way to do this seems undocumented at the time being https://www.apollographql.com/docs/apollo-server/data/file-uploads/ even if the option seems to be available when I read the source.

4. I had seemingly unrelated CORS errors popping up when I tried this : apparently I shouldn't have been sending `Authorization: null` header. I fixed it by not sending this header when I had no access-token : this has nothing to do with this PR per say, but I think it's likely people may have issues with CORS when opting in with this feature ? I still don't fully grasp how all this CORS thing work, and maybe the documentation should be improved. https://www.apollographql.com/docs/apollo-server/deployment/lambda/

5. I would also have appreciated having a direct link to the "client side" of things regarding file uploads in the "server side" documentation because even if I'm a backend developer, I wanted to create a micro client app to test everything was working correctly.

6. My use case was to transfer the received file to AWS S3. I tried providing the stream as is, however it failed because AWS S3 requires to send the `ContentLength` and couldn't determine the content length before consuming the stream. It would have greatly helped if I had had access to the size of the file in addition to its mimetype and encoding. It seems to me this information could be provided at no cost, but I might be wrong. I ended up providing it as a Buffer by creating this small util: 

```js
const streamToBuffer = stream =>
  new Promise(async resolve => {
    const chunks = []
    stream
      .on("data", data => chunks.push(data))
      .on("end", () => resolve(Buffer.concat(chunks)))
  })
```

I guess I have to open up an issue on https://github.com/jaydenseric/graphql-upload ?

7. `stream` property is now deprecated, and one should use `createReadStream()` instead: this is not documented correctly at the moment.

Overall : the code works really well. There are a number of loop holes in the documentation that could easily be fixed I guess. I'm willing to suggest some changes to the documentation through some PRs 👍 I can also create separate issues if needed.

**Comment by k00k** [comment]
@aymericbouzy Regarding some of the points that you highlighted, I had made an [example repository](https://github.com/abcya/apollo-server-lambda-example) that talked about the binary media types for API GW. It also has it all working with s3, were you not aware of this example? If you were and it wasn't helpful, please let me know why so I can make it more so.

File: packages/apollo-server-lambda/src/ApolloServer.ts [hunk 2]
```diff
@@ -213,19 +225,29 @@ export class ApolloServer extends ApolloServerBase {
         }
       }
```
Comment (abernix) [inline packages/apollo-server-lambda/src/ApolloServer.ts]: I'll make reference to this again in comments below, but is there a reason that most of the logic that's being added in `lambdaApollo.ts` (below, in this PR) isn't added here instead (before `graphqlLambda` is ever kicked off)?

This is more of what I was thinking with my comment on the other PR (https://github.com/apollographql/apollo-server/pull/1739/files#r288265931) and would make this match the other integrations.  This is most important from a maintainability standpoint over the the Apollo Server project as a whole since this divergent pattern makes it difficult to grok what's going on (particularly by starting a new pattern which passes _two_ option objects to `graphqlLambda` which already has an `options` property for its `GraphQLOptions`.

File: packages/apollo-server-lambda/src/lambdaApollo.ts [hunk 3]
```diff
@@ -39,36 +43,74 @@ export function graphqlLambda(
...
-      method: event.httpMethod,
-      options: options,
-      query:
-        event.httpMethod === 'POST' && event.body
-          ? JSON.parse(event.body)
-          : event.queryStringParameters,
-      request: {
-        url: event.path,
+
+    const runQuery = (query:any) => {
```
Comment (abernix) [inline packages/apollo-server-lambda/src/lambdaApollo.ts]: Would the type of this `query` be a `HttpQueryRequest`?: 

https://github.com/apollographql/apollo-server/blob/2728b9be5e1f84ea0abebcc97bb60082f82fd8ed/packages/apollo-server-core/src/runHttpQuery.ts#L24

**Comment by abernix** [comment]
I should add, most of all: I still think this is a valuable addition to the v2 series. If you’re willing to keep working through it, I’m happy to see it through!

File: packages/apollo-server-lambda/src/ApolloServer.ts (hunk 2 above)
Comment (k00k) [reply packages/apollo-server-lambda/src/ApolloServer.ts]: I believe the reason that everyone (me and the orig PR) has been organizing the code to process the files first, before doing the actual query is because the `processFileUploads` function's response is actually needed for the query body when it is a multipart form.  We're struggling to follow the same design principles of the other integrations (Express, HAPI, Fastify, etc) because those integrations all have the benefit of actual proper middleware implementations at the server level and that's where those implementations process the uploads. The lambda server/proxy doesn't really have that level of middleware as I understand it. And I'm pretty sure we're not going to get into a complete re-write of that part here. However, in those other integrations, middleware acts much the same way as I understand it, by handling the file upload first and then performing the query.

All that said, I did take another stab at it, moving the logic out to `ApolloServer.ts`. The `graphqlLambda` function kicks off the chain, and we kinda cheat by passing `processEvent` as the first arg so that we can await the file upload processing in `lambdaApollo.ts`. Not sure it really changes anything though, as we still wait for the upload to finish before running the query, but that seems to be the nature of the beast. If you'd like to see this WIP, you can look at the work in this branch: https://github.com/abcya/apollo-server/tree/event-rework/packages/apollo-server-lambda/src
Please ignore the `any` typings all over the place, was just trying to prove a concept quickly. Would definitely need some work there, just didn't want to put in too much effort without discussion (I'm not convinced myself).

The more time we spend on this, the more I see why the original PR structured it the way he did with the `.then()`, and I actually think it's a decent approach given the limitations we're up against.

**Comment by k00k** [comment]
@abernix any feedback on my [reply about organization of logic](https://github.com/apollographql/apollo-server/pull/3676#discussion_r379113970)?

**Comment by smschick** [comment]
@abernix any feedback on @k00k 's reply about organization of logic?   This is getting stagnant and we'd love to get this implemented... or at least get some closure.

**Comment by mathix420** [comment]
any news there ?

**Comment by erickrawczyk** [comment]
@abernix any way to help move this forward?

**Comment by abernix** [comment]
Superseded by https://github.com/apollographql/apollo-server/pull/3926, which has merged into `release-2.13.0`.  An _alpha_ release has been published on the `alpha` npm tag which includes #3926, and can be installed with `npm install apollo-server-lambda@2.13.0-alpha.0`.  Please try it out and report any issues on a new issue which references #3926.

**Comment by mateo2181** [comment]
> @jacky-ttt After some testing, I believe I was able to replicate the problem you were having. This is a guess, of course, because I don't know exactly all of the steps you did or didn't take. But I was able to upload locally and get "corrupted" files when sent through a real Lambda.
> 
> I'm guessing it's likely due to needing to set API Gateway's Binary Media Types. In order to pass binary data via `multipart/form-data`, you need to add that as a _Binary Media Type_ in your API Gateway's settings for that lambda.
> 
> I've published an [example repository](https://github.com/abcya/apollo-server-lambda-example) that takes care of this for you if you deploy with the [Serverless Framework](https://serverless.com/). Otherwise, the README outlines how to do it manually.
> 
> Please let me know if that does in fact solve your problem.
> 
> If anyone else has other/better ideas on how to handle that, please feel free to chime in. But I believe it's likely out of the scope of this repo and thus shouldn't affect the merging of this PR.

@k00k I followed your comment and follow the steps in your repo but I can't uplod images with my lambda function. I was working with serverless-offline I everything works fine, but I had not success with my function in AWS. 

Do you know which can be the problem?

Thanks in advance!

**Comment by k00k** [comment]
> @k00k I followed your comment and follow the steps in your repo but I can't uplod images with my lambda function. I was working with serverless-offline I everything works fine, but I had not success with my function in AWS.
> 
> Do you know which can be the problem?

If you did either the manual method or used the `serverless-apigw-binary ` plugin, that should do it. Not sure outside of that.

**Comment by mathix420** [comment]
> @k00k I followed your comment and follow the steps in your repo but I can't uplod images with my lambda function. I was working with serverless-offline I everything works fine, but I had not success with my function in AWS.
> 
> Do you know which can be the problem?
> 
> Thanks in advance!

It can be plenty of things did you watch your lambda logs ?  Can you provide us more details

**Comment by mateo2181** [comment]
Yes, I had an error connecting with Cloudinary. Now is working perfectly! 
Thanks!
